test/suites/auth.sh

```diff
@@ -199,6 +202,42 @@ effective_permissions: []"
   # The TLS identity is not trusted after deletion.
   [ "$(LXD_CONF="${LXD_CONF2}" lxc_remote query tls:/1.0 | jq -r '.auth')" = "untrusted" ]
 
+  # Check a TLS identity can update their own certificate.
+  # First create a new TLS identity and add it to test-group
+  LXD_CONF4=$(mktemp -d -p "${TEST_DIR}" XXX)
+  LXD_CONF="${LXD_CONF4}" gen_cert_and_key "client"
+  token="$(lxc auth identity create tls/test-user4 --quiet)"
+  LXD_CONF="${LXD_CONF4}" lxc_remote remote add tls "${token}"
+  lxc auth identity group add tls/test-user4 test-group
+
+  # Create another certificate to update to
+  LXD_CONF5=$(mktemp -d -p "${TEST_DIR}" XXX)
+  LXD_CONF="${LXD_CONF5}" gen_cert_and_key "client"
+
+  # We're using my_curl because the lxc wrapper function splits the --data argument on the spaces between "BEGIN CERTIFICATE" and lxc query returns a usage error.
+  # We could use lxc edit as it accepts stdin input, but replacing the certificate in the yaml was quite complicated.
+
+  # This asserts that test-user4 cannot change their own group membership
+  [ "$(LXD_CONF="${LXD_CONF4}" my_curl "https://${LXD_ADDR}/1.0/auth/identities/tls/test-user4" -X PUT --data "{\"tls_certificate\":\"$(awk '{printf "%s\\n", $0}' "${LXD_CONF5}/client.crt")\"}" | jq -r '.error_code')" -eq 403 ]
+
+  # This asserts that test-user4 can change their own certificate as long as the groups are unchanged
+  [ "$(LXD_CONF="${LXD_CONF4}" my_curl "https://${LXD_ADDR}/1.0/auth/identities/tls/test-user4" -X PUT --data "{\"tls_certificate\":\"$(awk '{printf "%s\\n", $0}' "${LXD_CONF5}/client.crt")\", \"groups\":[\"test-group\"]}" | jq -r '.status_code')" -eq 200 ]
+
+  # The original certificate is untrusted after the update
+  [ "$(LXD_CONF="${LXD_CONF4}" lxc_remote query tls:/1.0 | jq -r '.auth')" = "untrusted" ]
+
+  # Add the remote for the lxc config directory with the other certificates. No token needed as we're already trusted.
+  LXD_CONF="${LXD_CONF5}" lxc remote add tls "${LXD_ADDR}" --accept-certificate --auth-type tls
+  [ "$(LXD_CONF="${LXD_CONF5}" lxc_remote query tls:/1.0 | jq -r '.auth')" = "trusted" ]
+
+  # Do the same tests with patch. test-user4 cannot change their group membership
+  [ "$(LXD_CONF="${LXD_CONF5}" my_curl "https://${LXD_ADDR}/1.0/auth/identities/tls/test-user4" -X PATCH --data "{\"tls_certificate\":\"$(awk '{printf "%s\\n", $0}' "${LXD_CONF4}/client.crt")\", \"groups\":[\"new-group\"]}" | jq -r '.error_code')" -eq 403 ]
+
+  # Change the certificate back to the original, using patch. Here no groups are in the request, only the certificate.
+  [ "$(LXD_CONF="${LXD_CONF5}" my_curl "https://${LXD_ADDR}/1.0/auth/identities/tls/test-user4" -X PATCH --data "{\"tls_certificate\":\"$(awk '{printf "%s\\n", $0}' "${LXD_CONF4}/client.crt")\"}" | jq -r '.status_code')" -eq 200 ]
+  [ "$(LXD_CONF="${LXD_CONF4}" lxc_remote query tls:/1.0 | jq -r '.auth')" = "trusted" ]
+  [ "$(LXD_CONF="${LXD_CONF5}" lxc_remote query tls:/1.0 | jq -r '.auth')" = "untrusted" ]
+
```

Comment:
Just a compliment: thanks for these beautiful tests and comments :D

Reply:
Thanks :smile: